expo/expo
# [docs] Add internal link rewrite MDX plugin

**Edit:** simplification is done, a stacked PR containing all url rewrite and fixes is pending (see #10983).

# Why

- We don't have tooling to validate if internal links still work, e.g. referencing proper file or header
- When we switched internal MD(X) links to `next/link`, we encountered some issues related to the difference between MD(X) non-trailing-slash and Next trailing-slash behavior. e.g. `./path/` and `../path/` weren't working anymore.

This PR fixes both. 😄 

Link validation | VSCode linking
--- | ---
![mdx-link-validation](https://user-images.githubusercontent.com/1203991/98680294-7cf28300-2361-11eb-86c6-17ccb1a3f54a.gif) | ![mdx-vscode-linking](https://user-images.githubusercontent.com/1203991/98680299-7e23b000-2361-11eb-9904-6fe197513537.gif)

# How

This link rewriter only modifies internal links between the pages, in MDX only. Anchor or HTML links aren't modified because it's only implemented as remark plugin (not rehype, html transformer).

It allows us to reference files instead of absolute urls, allowing us to implement tooling around validation of these links. ~~I implemented it to be backwards compatible with our current links, although the tooling can't find those files and will warn that links aren't valid.~~

With this, we can write links like:

### From `/push-notifications/overview.md`, to `/push-notifications/using-fcm.md`
- `using-fcm` → link validation only works with `.md` reference
- ~~`../using-fcm/` → also link validation failure, but our current way of linking~~
- `../using-fcm.md` → link validation failure, hard-codes trailing-slash behavior, file isn't located `../`
- `./using-fcm.md` → fully picked up by auto validator
- `using-fcm.md` → fully picked up by auto validator
- `using-fcm.md#some-header` → fully picked up by auto validator

# Test Plan

- See unit tests for the rewriter
- See commits as paper-trail for the changes
- `$ yarn lint-links` to validate internal links (takes ±30s)
- `$ yarn dev`
	1. check the index page
	2. check the push-notifications overview page, first 3 links in list

# ~~TODO~~

~~Before merging this, I'd love to simplify the setup by updating the internal linking to follow 1 format. Preferably one that's supported by the `remark -u validate-links` command. With that, we can simplify the rewriter by _a lot_.~~

~~It's also weird that paths like `../../workflow/linking` works from `/versions/v39.0.0/sdk/app-auth.md`. It should link to `/versions/workflow/linking`, but the error page probably redirects to the proper url (`/workflow/linking`). That's also something I'd like to see fixed.~~
